proposals/NNNN-sequence-grouping.md

```diff
@@ -0,0 +1,249 @@
+# Sequence `grouped(by:)` and `keyed(by:)`
+
+* Proposal: [SE-NNNN](NNNN-filename.md)
+* Authors: [Alexander Momchilov](https://github.com/amomchilov)
+* Review Manager: TBD
+* Status: **Awaiting review**
+* Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN) or [apple/swift-evolution-staging#NNNNN](https://github.com/apple/swift-evolution-staging/pull/NNNNN)
+* Review: ([pitch](https://forums.swift.org/...))
+
+## Introduction
+
+This proposal would add new APIs on `Sequence` which let you group up or key elements in a more natural and fluent way than is currently possible with `Dictionary`'s initializers.
+
+## Motivation
+
+[SE-0165 Dictionary & Set Enhancements](https://github.com/apple/swift-evolution/blob/main/proposals/0165-dict.md) introduced some great utility APIs to Dictionary. Relating to this proposal are these 3 initializers:
+
+1. [`Dictionary.init(grouping:by:)`](https://developer.apple.com/documentation/swift/dictionary/init(grouping:by:))
+2. [`Dictionary.init(uniqueKeysWithValues:)`](https://developer.apple.com/documentation/swift/dictionary/init(uniquekeyswithvalues:))
+3. [`Dictionary.init(_:uniquingKeysWith:)`](https://developer.apple.com/documentation/swift/dictionary/init(_:uniquingkeyswith:))
+
+These APIs have proven very useful, and have replaced a plethora of hand-rolled calls to `reduce(into: [:]) { ... }`, but as initializers, they have some usability short comings.
+
+### Grouping
+
+It's not uncommon that you would like to group the results of a chain of method calls that transform a Sequence, and perhaps even keep chaining more transformations onto that Dictionary. For example:
+
+```swift
+let studentsWithDuplicateNames = Dictionary(
+		grouping: loadPeople()
+			.map { json in decoder.decode(Person.self, from: json) }
+			.filter(\.isStudent),
+		by: \.firstName
+	).filter { name, students in 1 < students.count }
```

Comment:
This boolean expression in this form is unintuitive, wouldn't it be better to put the count on the left hand side instead?
```suggestion
	).filter { name, students in students.count > 1 }
```

Reply:
Yeah I can understand, I usually put the value being tested on the left side and the value to test against on the right, and in this case in particular, the code reads more like what you would say out loud to describe the code.

Reply:
Ahhh you're right! If I slow down to read this as prose (and not just read it through quickly with a "code" perspective), `students.count > 1` does read like the English. I'll change this next time I'm making edits.